### PR TITLE
refactor: Implement timers as resources, take 2

### PR DESCRIPTION
**Pull request by bartlomieju**
Alternative to #3265 that keeps JS implementation of timeout and interval but changes Rust one. Instead of one global timer stored on `State`, each node of RBTree (representing one instant) has one timeout resource.

### CODE REVIEW

**Comment by bartlomieju** [comment]
CC @piscisaureus @ry

File: cli/js/timers.ts [hunk 1]
```diff
@@ -69,36 +50,37 @@ function schedule(timer: Timer, now: number): void {
...
   timer.scheduled = true;
-  // If the new timer is scheduled to fire before any timer that existed before,
-  // update the global timeout to reflect this.
-  if (globalTimeoutDue === null || globalTimeoutDue > timer.due) {
-    setOrClearGlobalTimeout(timer.due, now);
+
+  // If tree node has no rid that means that timeout wasn't yet scheduled in Rust - do it now
+  if (!dueNode.rid) {
+    dueNode.rid = sendSync(dispatch.OP_SET_TIMEOUT, { timeout });
+    pollTimeout(dueNode.rid, dueNode.due);
```
Comment (ry) [inline cli/js/timers.ts]: It seems that `setTimeout()` would now require two ops to be called... Perhaps not such a big deal, but it is less than currently (which is usually 1 op, but sometimes 0 ops if another timer is already pending)

File: cli/js/timers.ts [hunk 2]
```diff
@@ -18,49 +18,30 @@ interface Timer {
```
Comment (ry) [inline cli/js/timers.ts]: If we land this, there should be a paragraph at the top of this file explaining the algorithm.

File: cli/js/timers.ts (hunk 1 above)
Comment (bartlomieju) [reply cli/js/timers.ts]: It does, but only if there is no other timeout scheduled for that instant

**Comment by bartlomieju** [comment]
> @bartlomieju - I think you've done a good job with this patch - moving the global timer to resource based timers. That said, I was hoping this refactor would reduce the amount of code we would have to maintain. I didn't anticipate the problem of ordering timers with the same timeout - which seems to result in a lot of machinery on the JS side. I'm not sure this overall is an improvement since it increases the overhead per `setTimeout` call and does not significantly reduce code complexity.
> 
> @piscisaureus has thought about timers a bit - I'd be interested in his take on it...

Thanks! I actually wanted to reduce all of the junk in JS (see #3265) but as you said ordering is a big deal. @piscisaureus confirmed with Tokio team that currently there's no way in Tokio to ensure ordering of `Delay`s so that's our best bet for now I guess. 

As for code complexity I actually believe that this patch is better than before - there's single timeout per instant (granted, requires 2 ops to be called, but is that such a big overhead?), there's no more juggling of global timer and gives a clear way to remove all of the other junk once (if) Tokio supports ordering of `Delay`s.

**Comment by ry** [comment]
It would be nice if the Tokio Delay just had the semantic we needed. I wonder how hard it would be to wrap a Delay in Rust but provide moment handling...

**Comment by bartlomieju** [comment]
> It would be nice if the Tokio Delay just had the semantic we needed. I wonder how hard it would be to wrap a Delay in Rust but provide moment handling...

Just found there's [`DelayQueue`](https://docs.rs/tokio/0.1.22/tokio/timer/struct.DelayQueue.html) however it explicitly states that:
> Items are returned ordered by their expirations. Items that are configured to expire first will be returned first. There are no ordering guarantees for items configured to expire the same instant. Also note that delays are rounded to the closest millisecond.

So that doesn't get us anywhere near. I'll open an issue in Tokio repo asking if it's reasonable to expect that delay ordering will be implemented.

**Comment by bartlomieju** [comment]
Discussed offline; because this PR doesn't remove all of JS complexity and using single global timer is more optimal I'm closing this PR

**Pull request closed without merging**